lib/ansible/playbook/base.py

```diff
@@ -299,6 +300,64 @@ def validate(self, all_vars=None):
 
         self._validated = True
 
+    def _load_module_defaults(self, name, value):
+        # Ensure module_defaults is a list of dictionaries, the keys
+        # of which are static and fully resolved action and group names
+
+        if value is None:
+            return
+
+        if not isinstance(value, list):
+            value = [value]
+
+        validated_module_defaults = []
+        for defaults_dict in value:
+            if not isinstance(defaults_dict, dict):
+                raise AnsibleParserError(
+                    "The field 'module_defaults' is supposed to be a dictionary or list of dictionaries, "
+                    "the keys of which must be static action, module, or group names. Only the values may contain "
+                    "templates. For example: {'ping': \"{{ ping_defaults }}\"}"
+                )
+            validated_defaults_dict = {}
+            for defaults_entry, defaults in defaults_dict.items():
+                # module_defaults do not use the 'collections' keyword, so actions and
```

Comment:
Wishing for a time machine on this one to go back to 2.10- in the collections world, ideally, an unqualified action_group entry should be assumed to live in the defining collection, not "ansible.legacy" (which implies that there's a builtin redirect- made sense in 2.10, not so much now). Unfortunately, there are already collections in the wild that define action_groups relying on the `ansible.legacy`/`ansible.builtin` behavior.

I think it'd be OK to break the `ansible.legacy` part (in favor of `ansible.builtin`)- collections that actually want that behavior should be free to define the entries explicitly as `ansible.legacy` (but realistically, most probably shouldn't, and I'd be fine with saying "you can't"). But to keep 2.10-era collections working, we'd have to implement this change as "first try to resolve from the current collection, then fall back to `ansible.builtin`".

The main reason for this is to keep the places where a collection needs to hardcode its own name inside itself small.

Reply:
I'm not sure if I really understand what you're saying.

This applies to module_defaults in a playbook, not action_groups in collections. If a group/action in action_groups is not fully qualified, it is part of the collection still, and not 'ansible.legacy'. A group definition in action_groups is always part of its own ns.coll.

Adding 'ansible.legacy' to unqualified group names was intended to differentiate from opting into redirected-only actions
```
module_defaults:
  group/ansible.builtin.aws:  # redirected group only
    ...
```
and something that should maintain legacy behavior:
```
module_defaults:
  group/aws:  # legacy and redirected group
    ...
```

Since there's been a big push from docs to use fully qualified names for everything it didn't seem harmful to allow legacy behavior more explicitly with a fully qualified group name. I waffled over that though, and I could refactor _resolve_group to include 'ansible.legacy' actions for `group/aws` without adding support for `group/ansible.legacy.aws`. Do you want me to remove this?

I would prefer to not break 'ansible.legacy' actions though with the unqualified group name, which is what this bit is mainly for. A collection could include 'ansible.legacy' actions in a group, but I don't think they have any incentive or particular reason to do that. I don't think many people outside of core were even aware of ansible.legacy until recently. Calling `aws_s3` which resolves to a legacy override should continue working with `group/aws:`, but I see no reason that `group/amazon.aws.aws` should add support for it.

Reply:
Sorry, I put this comment in the wrong spot- the `module_defaults` side is fine as-is. My beef was with the `action_groups` plugin canonicalization in `mod_args.py` for action/module entries that are unqualified. Does what I said make sense if you re-evaluate it in that context? :laughing:

Reply:
I'm not sure, it's kind of part of the same train of thought. mod_args is also just for playbooks, not for resolving actions in action_groups. Here where you commented adds support for explicit FQ ansible.legacy group names and mod_args.py adds support for explicit FQ ansible.legacy action names. I'm not sure if one makes sense without the other. So whereas before this was required for legacy:
```
module_defaults:
  group/aws:
    ...
tasks:
  - aws_s3: # which resolves to legacy, not the collection
```
Now someone can update everything to use fully qualified if panicked by the FQ-or-you're-doing-it-wrong messaging implied by the docs:
```
module_defaults:
  group/ansible.legacy.aws:
    ...
tasks:
  - ansible.legacy.aws_s3:  # which resolves to legacy, not the collection
```
If I'm removing the mod_args stuff, then I think it would also make sense to remove the support for `group/ansible.legacy.<name>` from this PR.

Reply:
This is the bit that's resolving the action_groups if they aren't using fully qualified action/group names on line 417 in _resolve_group: 
```
            # The collection may or may not use the fully qualified name.
            # If not, it's part of the current collection.
```

Reply:
I removed the 'ansible.legacy' prefixing for groups and addressed the second part by removing the implicit ansible.legacy module inclusion and noting it in the changelog.

I'm going to rebase so I can update the second collection loader too since it blows up on the 'metadata' dict.